src/main/java/io/jenkins/plugins/gitlabbranchsource/GitLabWebHookListener.java

```diff
@@ -39,30 +40,45 @@ public void onNoteEvent(NoteEvent noteEvent) {
     @Override
     public void onMergeRequestEvent(MergeRequestEvent mrEvent) {
         LOGGER.log(Level.FINE, mrEvent.toString());
-        final long triggerDelay = findTriggerDelay(mrEvent.getProject().getWebUrl());
         GitLabMergeRequestSCMEvent trigger = new GitLabMergeRequestSCMEvent(mrEvent, origin);
-        SCMHeadEvent.fireLater(trigger, triggerDelay, TimeUnit.SECONDS);
+        fireTrigger(trigger, mrEvent.getProject().getWebUrl());
     }
 
     @Override
     public void onPushEvent(PushEvent pushEvent) {
         LOGGER.log(Level.FINE, pushEvent.toString());
-        final long triggerDelay = findTriggerDelay(pushEvent.getProject().getWebUrl());
         GitLabPushSCMEvent trigger = new GitLabPushSCMEvent(pushEvent, origin);
-        SCMHeadEvent.fireLater(trigger, triggerDelay, TimeUnit.SECONDS);
+        fireTrigger(trigger, pushEvent.getProject().getWebUrl());
     }
 
     @Override
     public void onTagPushEvent(TagPushEvent tagPushEvent) {
         LOGGER.log(Level.FINE, tagPushEvent.toString());
-        final long triggerDelay = findTriggerDelay(tagPushEvent.getProject().getWebUrl());
         GitLabTagPushSCMEvent trigger = new GitLabTagPushSCMEvent(tagPushEvent, origin);
-        SCMHeadEvent.fireLater(trigger, triggerDelay, TimeUnit.SECONDS);
+        fireTrigger(trigger, tagPushEvent.getProject().getWebUrl());
     }
 
-    private long findTriggerDelay(final String projectUrl) {
+    private void fireTrigger(final SCMHeadEvent<?> trigger, final String projectUrl) {
         final GitLabServer projectServer = findProjectServer(projectUrl);
+        if (findImmediateHookTrigger(projectServer)) {
+            SCMHeadEvent.fireNow(trigger);
```

Comment:
This should not fire the event twice? :)

```suggestion
            SCMHeadEvent.fireNow(trigger);
            return;
```

Reply:
That's the whole point of the MR, to fire it twice (if opted in through the server configutation).
Once immediately (that might hit the cache and do nothing), then a later one (default to 30 seconds) as a fallback in case the first one hit the cache (or do nothing if the changes are already being built).

If we wanted to fire it immediately and return we could set the value to 0 with the current version (fireLater with 0 seconds delay).